[sanitizer_symbolizer] Use correct format strings for uptr

DimitryAndric opened the pull request:
When compiling the common sanitizer libraries, there are many warnings about format specifiers, similar to:

    compiler-rt/lib/sanitizer_common/sanitizer_symbolizer_markup.cpp:31:32: warning: format specifies type 'void *' but the argument has type 'uptr' (aka 'unsigned long') [-Wformat]
       31 |   buffer->AppendF(kFormatData, DI->start);
          |                   ~~~~~~~~~~~  ^~~~~~~~~
    compiler-rt/lib/sanitizer_common/sanitizer_symbolizer_markup_constants.h:33:46: note: format string is defined here
       33 | constexpr const char *kFormatData = "{{{data:%p}}}";
          |                                              ^~
          |                                              %lu
    compiler-rt/lib/sanitizer_common/sanitizer_symbolizer_markup.cpp:46:43: warning: format specifies type 'void *' but the argument has type 'uptr' (aka 'unsigned long') [-Wformat]
       46 |   buffer->AppendF(kFormatFrame, frame_no, address);
          |                   ~~~~~~~~~~~~            ^~~~~~~
    compiler-rt/lib/sanitizer_common/sanitizer_symbolizer_markup_constants.h:36:48: note: format string is defined here
       36 | constexpr const char *kFormatFrame = "{{{bt:%u:%p}}}";
          |                                                ^~
          |                                                %lu
    ...

This is because `uptr` is dependent on the platform, and can be either `unsigned long long`, `unsigned long`, or `unsigned int`.

To fix the warnings, define the correct printf modifier for `uptr` and `sptr` as `SANITIZER_PTR_MOD`, and use it in the various format strings in `sanitizer_symbolizer_markup_constants.h`.

vitalybuka commented on compiler-rt/lib/sanitizer_common/sanitizer_symbolizer_markup_constants.h:
```diff
@@ -26,20 +26,23 @@ constexpr const char *kFormatDemangle = "{{{symbol:%s}}}";
 constexpr uptr kFormatDemangleMax = 1024;  // Arbitrary.
 
 // Function name or equivalent from PC location.
-constexpr const char *kFormatFunction = "{{{pc:%p}}}";
```
if uptr used for pointer, it should be `%p` but printf arg should be caster to (void*)
if it 's uptr for size, then `%zx` (for hex), or `%zu` should be used

DimitryAndric commented:
Casting seems uglier to me, though. On the other hand, it's difficult to make printf-like functions type-safe in any way...

DimitryAndric committed:
[sanitizer_symbolizer] Cast arguments for format strings in markup

When compiling the common sanitizer libraries, there are many warnings
about format specifiers, similar to:

    compiler-rt/lib/sanitizer_common/sanitizer_symbolizer_markup.cpp:31:32: warning: format specifies type 'void *' but the argument has type 'uptr' (aka 'unsigned long') [-Wformat]
       31 |   buffer->AppendF(kFormatData, DI->start);
          |                   ~~~~~~~~~~~  ^~~~~~~~~
    compiler-rt/lib/sanitizer_common/sanitizer_symbolizer_markup_constants.h:33:46: note: format string is defined here
       33 | constexpr const char *kFormatData = "{{{data:%p}}}";
          |                                              ^~
          |                                              %lu
    compiler-rt/lib/sanitizer_common/sanitizer_symbolizer_markup.cpp:46:43: warning: format specifies type 'void *' but the argument has type 'uptr' (aka 'unsigned long') [-Wformat]
       46 |   buffer->AppendF(kFormatFrame, frame_no, address);
          |                   ~~~~~~~~~~~~            ^~~~~~~
    compiler-rt/lib/sanitizer_common/sanitizer_symbolizer_markup_constants.h:36:48: note: format string is defined here
       36 | constexpr const char *kFormatFrame = "{{{bt:%u:%p}}}";
          |                                                ^~
          |                                                %lu
    ...

This is because `uptr` is dependent on the platform, and can be either
`unsigned long long`, `unsigned long`, or `unsigned int`.

To fix the warnings, cast the arguments to the expected type of the
format strings.

vitalybuka commented on compiler-rt/lib/sanitizer_common/sanitizer_symbolizer_markup.cpp:
```diff
@@ -96,7 +97,7 @@ static void RenderModule(InternalScopedString *buffer,
     buildIdBuffer.AppendF("%02x", module.uuid()[i]);
 
-  buffer->AppendF(kFormatModule, moduleId, module.full_name(),
+  buffer->AppendF(kFormatModule, static_cast<int>(moduleId), module.full_name(),
```
here, I'd rather update
"{{{module:%zu:%s:elf:%s}}}";

vitalybuka commented on compiler-rt/lib/sanitizer_common/sanitizer_symbolizer_markup.cpp:
```diff
@@ -118,8 +119,11 @@ static void RenderMmaps(InternalScopedString *buffer,
-                    accessBuffer.data(), range.beg - module.base_address());
+    buffer->AppendF(
+        kFormatMmap, reinterpret_cast<void *>(range.beg),
+        static_cast<unsigned int>(range.end - range.beg),
```
it's unlucky sanitizer convension, but uptr is equivalend to size_t.

`static_cast<unsigned int>(range.end - range.beg),` is uptr so size format `%z` should be used

So can you try to "{{{mmap:%p:0x%zx:load:%zu:%s:0x%zx}}}";

vitalybuka commented on compiler-rt/lib/sanitizer_common/sanitizer_symbolizer_markup.cpp:
```diff
@@ -119,11 +119,10 @@ static void RenderMmaps(InternalScopedString *buffer,
-        static_cast<int>(moduleId), accessBuffer.data(),
-        static_cast<unsigned int>(range.beg - module.base_address()));
+    buffer->AppendF(kFormatMmap, reinterpret_cast<void *>(range.beg),
+                    static_cast<size_t>(range.end - range.beg),
```
no need to cast size_t

I am saying that compiler-rt (sanitizers) use uptr to format as size_t.
It's all over the code, it should be OK to pass uptr into %zx.

DimitryAndric committed:
Also adjust format string for moduleId.

vitalybuka commented:
Thank you!
